### ISSUE TITLE
Setting up insta360 x4 with macos & NVIDIA jetson device,

### ISSUE THREAD

**Issue by autonomy-123** [Insta360Develop/CameraSDK-Cpp#65]
Hi everyone,
I am working with nvidia jetson & want to use insta 360 x4  with it so that  i can see live feed on the jetson device system & able to capture the images & videos on to the local system itself.

I am trying to connect to jeston with usb but it says unable to mount disk...
On macOS able to connect with wifi but not able see live feed etc.

Anyone have idea how to achieve this task. step by step guide would be grateful.

Thanks

**Comment by autonomy-123** [comment]
> Hi everyone, 
@le-insta360 @capjason-insta360 @capjason 

I am working with nvidia jetson & want to use insta 360 x4 with it so that i can see live feed on the jetson device system & able to capture the images & videos on to the local system itself.
> 
> I am trying to connect to jeston with usb but it says unable to mount disk... On macOS able to connect with wifi but not able see live feed etc.
> 
> Anyone have idea how to achieve this task. step by step guide would be grateful.
> 
> After connecting insta360 x4 with usb & android mode on, installing libusb-dev, libudev-dev depenedencies, I am running main.cc file under demo folder. It throughs error
> 
> ![WhatsApp Image 2024-11-11 at 20 13 16](https://private-user-images.githubusercontent.com/187417997/384962655-bbe6b6fc-efa3-4d36-8908-8565eacfdca2.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE0MDY5NDksIm5iZiI6MTczMTQwNjY0OSwicGF0aCI6Ii8xODc0MTc5OTcvMzg0OTYyNjU1LWJiZTZiNmZjLWVmYTMtNGQzNi04OTA4LTg1NjVlYWNmZGNhMi5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTExMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMTJUMTAxNzI5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmFlMTI2YmU4NjZlZjljOTNhMjFhNGY4MTVkNGIyMzQ0ZDBjZDdmYTY4NTU1MzYyYmRlYzE4YmJiYmRhMDUxMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.2VBq9H8EihjQyZHdC7x8QR7Mv6gMbaMS6IA-b-fOu58)
> 
> Let me know how I can achieve my task.
> 
> Thanks

**Comment by MaxDu17** [comment]
Hi Moksh, how are you planning to run the SDK on Jetson, given that it's an ARM processor and the .so is compiled for X86? I'm trying to do the same thing on Jetson and that's the problem I'm running into--let me know if you're able to make progress!

**Comment by Moksh-Mehta-1937** [comment]
Really depends on your use case and what you are trying to do. I encountered the above error on my Linux machine (X86) just when I started to explore the SDK (a beginner mistake)

My goal is to run detection algorithms on the camera's feed so rather than using the SDK I can just switch the camera to Webcam Mode and use OpenCV in Python to get the raw frames. Once I get that, I can basically do anything.

I realized it the hard way that over-engineering is not always a good thing. Let me know what your use-case is and perhaps I can help? As for using the SDK on Jetson there are some methods you can explore:

1.) There is an SDK for Android as well available to download, I am not sure if it is possible but the Android SDK might have .so files compatible with the ARM architecture, you can do some experiments to see if it works. Of-course, if this works you might have to create a wrapper to call functions in your desired coding language

2.) Best method is to use it as a webcam. I know certain functions (that the SDK provides) will not be available but you can always find a workaround to make it work for your use case and I would suggest you try with this method.

3.) If you have to absolutely run the SDK on Jetson, you can try an emulator. However, based on my past experience, this approach is highly impractical. I strongly recommend against spending time on it.

I have also reached out to Insta360 to ask for an aarch64 compatible version of the SDK and am still waiting for a reply. Also, I am not that courageous to spend time trying to cross-compile my pipeline when I know it is not going to work.

**Comment by MaxDu17** [comment]
Yes I agree that webcam mode can work in a lot of applications, although my particular application requires access to the raw fisheye feeds, which unfortunately the insta360 only provides through the SDK. I've also reached out to Insta360. I'll look into emulators like Qemu or even a docker container, and I'll update this thread if I find something that works. Given that many portable systems use ARM, I wish that Insta360 would just provide an arm-compiled version of the binary upfront.

**Comment by Moksh-Mehta-1937** [comment]
If you specifically need the raw fisheye feed using the SDK on a Jetson device, getting it to work will be extremely challenging. In fact, I’m not even sure if QEMU would be a viable solution. (BTW were you able to run it on your local X86 machine?)

One workaround is to use an intermediate relay device, something similar to a Raspberry Pi, that can communicate with both the Insta360 X4 and the Jetson. Suitable options could include a ZimaBoard or a LattePanda 3, both of which have detailed documentation available online for further research (and there are many more out there).

The role of the relay device would be to capture the camera feed, receiving raw H.264 or H.265 packets through the SDK, and then transmit those packets to the Jetson device, either via a wired or wireless connection.

This was my initial plan and although it is quite a task, it is fairly easy compared to figuring out how to run X86 programs on ARM devices. I am not sure about the budget and cost of these boards but I assume its going to cost some dollars.

**Comment by Double-Insta360** [comment]
If you still need the Jetson version, you can contact me at xuyongbo@insta360.com, but please note that we rarely maintain or support this version.
